hilo/hilo_generator.go

```diff
@@ -0,0 +1,277 @@
+        return fmt.Sprintf(`%s%s-%d`,ref.prefix, next_id, ref.server_tag)
+}
+/// Generates the document ID.
+func (ref *HiLoKeyGenerator) generateDocumentKey() string {
+		// lock this until not change range currently
+        ref.lock.Lock()
+        defer ref.lock.Unlock()
+        ref.collection_ranges[ref.tag] = RangeValueDefault()
+        ref.next_id()
+        return ref.getDocumentKeyFromId(ref.collection_ranges[ref.tag].current)
+}
+func (ref *HiLoKeyGenerator) next_id() {
+		// todo: убей не пойму - зачем здесь цикл , если мы блокируем все выше
+		//for  true {
+            my_collection_range, ok := ref.collection_ranges[ref.tag]
+            if !ok {
+                return
+            }
+            my_collection_range.current ++
+            if my_collection_range.current <= my_collection_range.max_id {
+                return
+            }
+                //// with ref.lock:
+                //if my_collection_range != ref.collection_ranges[ref.tag] {
+                //    continue
+                //}
+                ref.getNextRange()
+                ref.collection_ranges[ref.tag] = ref.rangeValues
+                // exceptions.FetchConcurrencyException
+            //}
+        //}
+                     
+}
+//todo: не стал описывать парсинг структуры, возможно, это вообще лишнее и стоит заменить на просто чтение прямо из тела ответа??
+type resHiloCommand struct {
+    prefix, serverTag, last_range_at string
+    low, high, last_size  uint
+}
+func (ref *HiLoKeyGenerator) getNextRange() error {
+    hilo_command, err := NewNextHiLoCommand(ref.tag, ref.last_batch_size, ref.last_range_at, ref.parent.parent.conventions.IdentityPartsSeparator, ref.rangeValues.max_id)
+    if err != nil {
+        return err
+    }
+    executor, err := ref.parent.parent.createExecutor()
+    if err == nil {
+        buf, err := executor.ExecuteOnCurrentNode(hilo_command, true)
+        if err != nil {
+            return err
+        }
+        var result resHiloCommand
+        err = json.Unmarshal(buf, result)
+        if err != nil {
+            return err
+        }
+        ref.prefix = result.prefix
+        ref.server_tag = result.serverTag
+        //todoL возможно, ято здесь сервер будет сразу отдавать в нужном формате и ненужно преоборазование?
+        ref.last_range_at, err = time.Parse("2006-01-02 15:04:05 -0700 MST", result.last_range_at)
+        if err != nil {
+            return err
+        }
+        ref.last_batch_size = result.last_size
+        ref.rangeValues = NewRangeValue(result.low, result.high)
+    }
+
+    return err
+}
+func (ref *HiLoKeyGenerator) returnUnusedRange() error {
+        return_command, err := NewHiLoReturnCommand(ref.tag, ref.rangeValues.current, ref.rangeValues.max_id)
+        if err != nil {
+            return err
+        }
+        //ref.store.GetRequestExecutor("")
+    executor, err := ref.parent.parent.createExecutor()
```

Comment:
if calling `createExecutor` you are creating a new one?
If so, that is a bad idea, assuming this will generate new HTTP connection.
We want to use connection pooling as much as possible, and the executor also handles failover so we don't want too many of theses.